src/BoundsErrorLight.jl

```diff
@@ -0,0 +1,60 @@
+"""
+    BoundsErrorLight <: Exception
+
+Like `BoundsError`, but doesn't store the array, to prevent clashing with escape analysis.
+
+Stores the type of the array and the shape instead.
+"""
+struct BoundsErrorLight <: Exception
+    type_of_array::DataType
+    shape::Tuple{Vararg{Int}}
+    index::Int
+end
+
+function show_index(io::IO, x)
+    if x isa Colon
+        print(io, ':')
+    elseif x isa Base.OneTo
+        print(io, "1:")
+        show(io, last(x))
+    else
+        show(io, x)
+    end
+    nothing
+end
+
+function Base.showerror(io::IO, ex::BoundsErrorLight)
+    typ = ex.type_of_array
+    shp = ex.shape
+    ind = ex.index
+    print(io, "BoundsErrorLight: attempt to access array of type ")
+    show(io, typ)
+    print(io, " and shape ")
+    show(io, shp)
+    print(io, " at index [")
+    if (ind isa AbstractRange) || (ind isa AbstractString)
+        show(io, ind)
+    else
+        for (i, x) ∈ enumerate(ind)
+            if 1 < i
+                print(io, ", ")
+            end
+            show_index(io, x)
+        end
+    end
+    print(io, ']')
+    nothing
+end
+
+# inline everything to help escape analysis
+
+function throw_boundserrorlight(A::AbstractArray, i::Int)
+    @inline
+    @inline throw(BoundsErrorLight(typeof(A), size(A), i))
+end
+
+function check_bounds_light(A::AbstractArray, i::Int)
+    @inline
+    @inline checkbounds(Bool, A, i) || @inline throw_boundserrorlight(A, i)
```

Comment:
Do we actually need to inline `throw` and `checkbounds`?  I still don't see memory allocations in https://github.com/JuliaArrays/FixedSizeArrays.jl/pull/116#issuecomment-2901913689 (and also in `@code_llvm f(per)`) if I remove those annotations.  That'd alleviate my concern of inlining too much.

Reply:
That'd also be closer to how `checkbounds` is defined for `AbstractArray`: https://github.com/JuliaLang/julia/blob/c5df01807e3b3549b2669afdea11526b80049f3a/base/abstractarray.jl#L695

Reply:
you probably can remove it.

Reply:
With this change
```suggestion
    throw(BoundsErrorLight(typeof(A), size(A), i))
end

function check_bounds_light(A::AbstractArray, i::Int)
    @inline
    checkbounds(Bool, A, i) || throw_boundserrorlight(A, i)
```
this would be good to go for me.